src/execution/operator/persistent/physical_delete.cpp

```diff
@@ -12,28 +13,69 @@ namespace duckdb {
 //===--------------------------------------------------------------------===//
 class DeleteGlobalState : public GlobalSinkState {
 public:
-	DeleteGlobalState() : deleted_count(0) {
+	DeleteGlobalState() : deleted_count(0), returned_chunk_count(0) {
 	}
 
 	mutex delete_lock;
 	idx_t deleted_count;
+	ChunkCollection return_chunk_collection;
+	idx_t returned_chunk_count;
+};
+
+class DeleteLocalState : public LocalSinkState {
+public:
+	DeleteLocalState(const vector<LogicalType> &table_types) {
+		delete_chunk.Initialize(table_types);
+	}
+	DataChunk delete_chunk;
 };
 
 SinkResultType PhysicalDelete::Sink(ExecutionContext &context, GlobalSinkState &state, LocalSinkState &lstate,
                                     DataChunk &input) const {
 	auto &gstate = (DeleteGlobalState &)state;
+	auto &ustate = (DeleteLocalState &)lstate;
+
+	// get rows and
+	auto &transaction = Transaction::GetTransaction(context.client);
+	auto &row_identifiers = input.data[row_id_index];
+
+	vector<column_t> column_ids;
+	for (idx_t i = 0; i < table.column_definitions.size(); i++) {
+		column_ids.emplace_back(i);
+	};
+	auto cfs = ColumnFetchState();
 
-	// delete data in the base table
-	// the row ids are given to us as the last column of the child chunk
 	lock_guard<mutex> delete_guard(gstate.delete_lock);
-	gstate.deleted_count += table.Delete(tableref, context.client, input.data[row_id_index], input.size());
+	if (return_chunk) {
+		if (row_identifiers.GetVectorType() == VectorType::FLAT_VECTOR ||
```

Comment:
Instead of checking the vector type, you can use `Normalify` to turn the `row_identifiers` into a flat vector. That should simplify the code here.